MapleServer2/Types/FieldNavigator.cs

```diff
@@ -0,0 +1,244 @@
+using Maple2.PathEngine.Types;
+using Maple2Storage.Types;
+using Maple2Storage.Types.Metadata;
+using Path = Maple2.PathEngine.Path;
+
+namespace MapleServer2.Types;
+
+public class FieldNavigator : IDisposable
+{
+    private readonly Mesh Mesh;
+    private readonly CollisionContext CollisionContext;
+
+    public FieldNavigator(string mapName)
+    {
+        byte[] meshBuffer = File.ReadAllBytes(Paths.NAVMESH_DIR + $"/{mapName}.tok");
+
+        Mesh = MapleServer.PathEngine.loadMeshFromBuffer(FileFormat.tok, meshBuffer, options: null);
+        CollisionContext = Mesh.newContext();
+    }
+
+    /// <summary>
+    /// Find a random position on the mesh around the agent with the given radius.
+    /// Creates a path from the agent position to the random position.
+    /// </summary>
+    /// <returns>List of CoordS or null if path is not possible</returns>
+    public List<CoordS> GenerateRandomPathAroundAgent(Agent agent, int radius)
+    {
+        Position randomPositionLocally = Mesh.generateRandomPositionLocally(agent.getPosition(), radius);
+
+        if (!Mesh.positionIsValid(randomPositionLocally))
+        {
+            return null;
+        }
+
+        using Path path = agent.findShortestPathTo(CollisionContext, randomPositionLocally);
+        return PathToCoordS(path);
+    }
+
+    /// <summary>
+    /// Find a random position on the mesh around the centerCoord with the given radius.
+    /// Creates a path from the centerCoord to the random position.
+    /// </summary>
+    /// <returns>List of CoordS or null if path is not possible</returns>
+    public List<CoordS> GenerateRandomPathAroundCoord(Agent agent, CoordS centerCoord, int radius)
+    {
+        Position position = FindPositionFromCoordS(centerCoord);
+
+        if (!Mesh.positionIsValid(position))
+        {
+            return null;
+        }
+
+        Position randomPositionLocally = Mesh.generateRandomPositionLocally(position, radius);
+
+        if (!Mesh.positionIsValid(randomPositionLocally))
+        {
+            return null;
+        }
+
+        using Path path = agent.findShortestPathTo(CollisionContext, randomPositionLocally);
+        return PathToCoordS(path);
+    }
+
+    /// <summary>
+    /// Find the shortest curved path from the agent to the target position.
+    /// </summary>
+    /// <returns>List of CoordS or null if path is not possible</returns>
+    public List<CoordS> FindPath(Agent agent, CoordS endCoord)
+    {
+        Position end = FindPositionFromCoordS(endCoord);
+
+        if (!Mesh.positionIsValid(end))
+        {
+            return null;
+        }
+
+        using Path shortestPath = agent.findShortestPathTo(CollisionContext, end);
+        if (shortestPath is null)
+        {
+            Console.WriteLine("Shortest path is null");
+            return null;
+        }
+
+        using Path path = agent.generateCurvedPath(shortestPath, CollisionContext, 0, 0, sectionLength: 50, turnRatio1: 0.5f, turnRatio2: 0.9f);
```

Comment:
I'm not sure how well a curved path will work given that a packet can only move npcs in a straight line.